Db refactor

jagerman opened the pull request:
Completely overhauls the underlying database design (see the added src/schema.sql):

- referential integrity
- data normalization
- using triggers for various updates to reduce the amount of data logic in code.
- indexes
- one single database rather than 1 + 1/room
- various structural improvements for useful features such as:
    - individually and default-assignable read/write/upload
    - time-based permission expiries (e.g. restrict this permission for x amount of time)
    - distinguishing between admins (who can add/remove moderators) and moderators (who can only moderate but not control the moderator list)
    - global server moderators/admins
    - global server bans
    - single VIEW for simple querying of a user's effective permissions (i.e. with database-side coalescing of global/local permissions).
    - pinned message via foreign key
    - room icon as a regular uploaded file
    - per-file file expiries (so that admins/mods can upload non-expiry files, such as the room icon, or pinned downloads).
    - message history so that when messages get edited or deleted a log is kept of the old value (the idea being that mods would be able to look this up).
    - movable messages (e.g. so there could be a "bad posts" room that only moderators have access to).
    
Still a bunch TODO on this:
- [x] existing queries need to be rewritten to query the new structure
- [x] migration from the multi table -> single table
- ~~exposing endpoints for potential new functionality~~ will come in a later PR.

jagerman committed:
Fix postinst to properly chown privkeys

jagerman committed:
Initial DB redesign

Completely overhauls the underlying database design:

- referential integrity
- data normalization
- using triggers for various updates to reduce the amount of data logic
  in code.
- indexes
- one single database rather than 1 + 1/room
- various structural improvements for useful features such as:
    - individually and default-assignable read/write/upload
    - time-based permission expiries (e.g. restrict this permission for
      x amount of time)
    - distinguishing between admins (who can add/remove moderators) and
      moderators (who can only moderate but not control the moderator
      list)
    - global server moderators/admins
    - global server bans
    - single VIEW for simple querying of a user's effective permissions
      (i.e. with database-side coalescing of global/local permissions).
    - pinned message via foreign key
    - room icon as a regular uploaded file
    - per-file file expiries (so that admins/mods can upload non-expiry
      files, such as the room icon, or pinned downloads).
    - message history so that when messages get edited or deleted a log
      is kept of the old value (the idea being that mods would be able
      to look this up).
    - movable messages (e.g. so there could be a "bad posts" room that
      only moderators have access to).

msgmaxim commented on src/handlers.rs:
```diff
@@ -575,14 +555,11 @@ pub fn get_messages(
     // Query the database
-    let raw_query: String;
+    let raw_query: &str;
     if query_params.get("from_server_id").is_some() {
```
Not sure if you are aware, but if statements are also expressions in Rust, so you could write:
```rust
let raw_query = if query_params.get("from_server_id").is_some() {
   "SELECT id, public_key, timestamp, data, signature FROM messages WHERE id > (?1) AND is_deleted = 0 ORDER BY id ASC LIMIT (?2)"
} else {
    "SELECT id, public_key, timestamp, data, signature FROM messages WHERE is_deleted = 0 ORDER BY id DESC LIMIT (?2)"
}
```
(if you prefer)

jagerman replied:
Thanks, appreciate any tips like this as I'm a rust noob.

msgmaxim commented on src/storage.rs:
```diff
@@ -1,365 +1,92 @@
-        Err(_) => return,
+    let conn = match DB_POOL.get() {
+        Ok(conn) => conn,
+        Err(e) => return error!("Couldn't prune files: {}.", e)
```
I realise that this wasn't introduced in this PR but worth improving this since it is being touched:
```suggestion
        Err(e) => {
            error!("Couldn't prune files: {}.", e);
            return;
        }
```
Otherwise it looks like we are "returning" the error which isn't the case...

jagerman replied:
I grabbed that from other code but yeah, I agree.

jagerman committed:
Apply arrived-at future user permissions

jagerman committed:
Rename identifier -> token, make case insensitive, add constraint

jagerman committed:
Fix grammatically incorrect 'due to'

"due to" works as an adjective that follows a noun (not verbs/verb
phrases) and is being used incorrectly in all the error messages fixed
here.

jagerman committed:
Database redesign and code associated refactor

jagerman committed:
Add --log-level; make --help messages more descriptive

jagerman committed:
Add shm/wal db files to .gitignore

jagerman committed:
Make it work

jagerman committed:
Fix file upload

jagerman committed:
Fix deletions; global moderator

- "Deleted" messages should be set to null, not actually deleted
- Add automatic updating of "edited" to the message update trigger
- Accomodate janky deprecated Session compact_poll logic that expects
  deletion lists even if it doesn't ask for them.
- Add a useful view for (admin direct sqlite) message querying.
- Fix message_details VIEW query

jagerman committed:
Fix room image expiry, fix file pruning

Uploaded room images shouldn't expire; this makes the expiry column
nullable and sets their expiry to null, and adds a trigger to expire
replaced room images.

Also fixes a panic in file deletion that made it not actually delete
files.

msgmaxim commented on src/rpc.rs:
```diff
@@ -1,105 +1,184 @@
+        user = Some(handlers::insert_or_update_user(&*storage::get_conn()?, &sessionid)?);
+
+        // Check for a global ban, and if so, terminate the request right away.
+        if user.as_ref().unwrap().banned {
```
Could you not check if `user` is banned before wrapping it in `Some`? I can see that it is safe to unwrap in this case, but it does add a bit of extra friction when reading the code.

jagerman committed:
Add stubs/fixmes for new endpoints

jagerman committed:
Add update counter on metadata; add room description

jagerman committed:
Don't store padding in db

Trim padding when storing and repad on the fly when retrieving.

jagerman committed:
Fix room user counts

jagerman committed:
Fix metadata trigger queries

They have to be defined after the tables on which they run, of course.
Also fixed some syntax errors in them and made them work.

jagerman committed:
Reformatting: cargo fmt + various manual tweaks

jagerman committed:
Add 0.1.x database migration

msgmaxim commented on src/handlers.rs:
```diff
@@ -994,6 +990,37 @@ pub fn get_member_count_since(user: User, room: Room, ago: Duration) -> Result<R
+/// since the given update value (or always, if since_update is omitted).
+pub fn poll_room_metadata(user: &User, room: Room, since_update: Option<i64>) {
+    // FIXME TODO
+    panic!("FIXME TODO");
```
There is a macro for this: `todo!() or todo!("message")`

darcys22 commented on src/handlers.rs:
```diff
@@ -1,1139 +1,1406 @@
+//
+// (The same applies to get_deleted_messages, below).
+
+fn get_messages_params(query_params: &HashMap<String, String>) -> (Option<i64>, u16) {
```
Telegram looks like they have 3 methods of retrieving messages:
- Limit provided as an arg from the client
- Offset skips a number of elements since beginning of list
- Offset from ID calls like loading 20 messages, older than message with ID MSGID: `messages.getHistory({offset_id: MSGID, add_offset: 0, limit: 20})`

https://core.telegram.org/api/offsets

jagerman committed:
Add missing description field to Room struct

jagerman committed:
Separate old Room response from new Room response

The old Room response included the room token as "id", which is quite
awkward, and not something we want to propagate into the future API, so
we can't just start returning the extra details if we don't want to
maintain that misnaming.

This changes the API so that existing (deprecated) endpoints get the
deprecated field, and the new Room deserialization is free to break
backwards compat.

jagerman committed:
fix 1

jagerman committed:
Support multiple pinned messages & related

- Rename `info_updated` -> `info_updates` to be consistent with `updates`.
- Relax triggers to only update `info_updates` instead of both updates
  and info_updates, and fix some trigger statements that were updating
  the wrong value.
- Add table for tracking (multiple) pinned messages.
- Trigger info_updates update when pinning/unpinning.

darcys22 committed:
test suite and cargp fmt

jagerman committed:
Add drone builds

darcys22 committed:
minor review notes
